core/services/head_tracker.go

```diff
@@ -104,103 +88,67 @@ func (ht *HeadTracker) logger() *logger.Logger {
+				"blockNumber", highestSeenHead.Number,
+				"blockHash", highestSeenHead.Hash,
+			)
+		}
 
-	ht.wgDone.Add(3)
-	go ht.listenForNewHeads()
-	go ht.backfiller()
-	go ht.headSampler()
+		ht.wgDone.Add(3)
+		go ht.headListener.ListenForNewHeads(ht.handleNewHead, ht.handleConnected)
+		go ht.backfiller()
+		go ht.headSampler()
 
-	ht.started = true
-	return nil
+		return nil
+	})
 }
 
 // Stop unsubscribes all connections and fires Disconnect.
 func (ht *HeadTracker) Stop() error {
-	ht.headMutex.Lock()
-
-	if !ht.started {
-		ht.headMutex.Unlock()
+	return ht.StopOnce("HeadTracker", func() error {
+		ht.logger().Info(fmt.Sprintf("HeadTracker disconnecting from %v", ht.store.Config.EthereumURL()))
+		close(ht.chStop)
+		ht.wgDone.Wait()
 		return nil
-	}
-
-	ht.logger().Info(fmt.Sprintf("Head tracker disconnecting from %v", ht.store.Config.EthereumURL()))
-	close(ht.done)
-	close(ht.subscriptionSucceeded)
-	ht.started = false
-	ht.headMutex.Unlock()
-
-	ht.wgDone.Wait()
-	return nil
+	})
 }
 
-// Save updates the latest block number, if indeed the latest, and persists
-// this number in case of reboot. Thread safe.
 func (ht *HeadTracker) Save(ctx context.Context, h models.Head) error {
-	ht.headMutex.Lock()
-	if h.GreaterThan(ht.highestSeenHead) {
-		ht.highestSeenHead = &h
-	}
-	ht.headMutex.Unlock()
-
-	err := ht.store.IdempotentInsertHead(ctx, h)
-	if ctx.Err() != nil {
-		return nil
-	} else if err != nil {
-		return err
-	}
-	return ht.store.TrimOldHeads(ctx, ht.store.Config.EthHeadTrackerHistoryDepth())
+	return ht.headSaver.Save(ctx, h)
 }
 
-// HighestSeenHead returns the block header with the highest number that has been seen, or nil
 func (ht *HeadTracker) HighestSeenHead() *models.Head {
-	ht.headMutex.RLock()
-	defer ht.headMutex.RUnlock()
+	return ht.headSaver.HighestSeenHead()
+}
 
-	if ht.highestSeenHead == nil {
-		return nil
-	}
-	h := *ht.highestSeenHead
-	return &h
+func (ht *HeadTracker) HighestSeenHeadFromDB() (*models.Head, error) {
+	return ht.headSaver.HighestSeenHeadFromDB()
+}
+
+func (ht *HeadTracker) Chain(ctx context.Context, hash common.Hash, depth uint) (models.Head, error) {
+	return ht.headSaver.Chain(ctx, hash, depth)
 }
 
 // Connected returns whether or not this HeadTracker is connected.
 func (ht *HeadTracker) Connected() bool {
-	ht.headMutex.RLock()
-	defer ht.headMutex.RUnlock()
-
-	return ht.connected
+	return ht.headListener.Connected()
 }
 
-func (ht *HeadTracker) connect(bn *models.Head) {
-	for _, trackable := range ht.callbacks {
-		if err := trackable.Connect(bn); err != nil {
-			ht.logger().Warn(trackable.Connect(bn))
-		}
-	}
+func (ht *HeadTracker) handleConnected() {
+	ht.connect(ht.headSaver.HighestSeenHead())
 }
 
-func (ht *HeadTracker) disconnect() {
-	for _, trackable := range ht.callbacks {
-		trackable.Disconnect()
+func (ht *HeadTracker) connect(bn *models.Head) {
```

Comment:
I look forward to dumping jpv1 so this connect callback can be destroyed forever :)

Reply:
ha, it's also used in BalanceMonitor :)
```
	// Connect head can be out of date, so always query the latest balance
	bm.checkBalance(nil)
```